File: java/src/main/java/com/camunda/academy/TestConnection.java
package com.camunda.academy;

import io.camunda.client.CamundaClient;
import io.camunda.client.api.response.Topology;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletionException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Camunda 8 SaaS Connection Test Tool - Java Version
 * 
 * This tool tests connectivity to Camunda 8 SaaS by:
 * 1. Checking Java version (must be 17+)
 * 2. Loading environment variables from envVarsExtended.txt
 * 3. Creating Camunda cloud client
 * 4. Testing connection via topology request
 * 
 * Exit codes:
 * 0 = Success
 * 1 = SSL Error
 * 2 = Connection Error
 * 3 = Authentication Error
 * 4 = Other/Unexpected Error
 * 
 * Author: Camunda Academy
 */
public class TestConnection {

    private static final int EXIT_SUCCESS = 0;
    private static final int EXIT_SSL_ERROR = 1;
    private static final int EXIT_CONNECTION_ERROR = 2;
    private static final int EXIT_AUTH_ERROR = 3;
    private static final int EXIT_OTHER_ERROR = 4;

    private static final String ENV_FILE = "envVarsExtended.txt";

    public static void main(String[] args) {
        try {
            checkJavaVersion();

            System.out.println("Loading environment variables...");
            Map<String, String> envVars = loadEnvVars();

            System.out.println("Connecting to Camunda 8 SaaS...");
            testConnection(envVars);

            printSuccess();
            System.exit(EXIT_SUCCESS);

        } catch (Exception e) {
            handleException(e);
        }
    }

    private static void checkJavaVersion() {
        String javaVersion = System.getProperty("java.version");
        String[] versionParts = javaVersion.split("\\.");
        int majorVersion;

        try {
            if (versionParts[0].equals("1")) {
                majorVersion = Integer.parseInt(versionParts[1]);
            } else {
                majorVersion = Integer.parseInt(versionParts[0]);
            }

            if (majorVersion < 17) {
                printError("Java 17 or higher is required. Current version: " + javaVersion +
                        "\nPlease upgrade Java or contact your training manager." +
                        "\nDownload Java from: https://adoptium.net/");
                System.exit(EXIT_OTHER_ERROR);
            }

            System.out.println("Java version check passed: " + javaVersion);

        } catch (NumberFormatException e) {
            System.out.println("Warning: Could not parse Java version, continuing anyway...");
        }
    }

    private static Map<String, String> loadEnvVars() throws IOException {
        Path envPath = Paths.get(ENV_FILE);

        if (!Files.exists(envPath)) {
            Path parentPath = Paths.get("..", ENV_FILE);
            if (Files.exists(parentPath)) {
                envPath = parentPath;
            } else {
                throw new IOException(ENV_FILE
                        + " file not found. Double check that this file is available in this directory or parent directory");
            }
        }

        Map<String, String> envVars = new HashMap<>();
        Pattern pattern = Pattern.compile("export\\s+(\\w+)='(.+)'");

        try (BufferedReader reader = new BufferedReader(new FileReader(envPath.toFile()))) {
            String line;
            int lineNum = 0;

            while ((line = reader.readLine()) != null) {
                lineNum++;
                line = line.trim();

                if (line.isEmpty() || line.startsWith("#")) {
                    continue;
                }

                Matcher matcher = pattern.matcher(line);
                if (matcher.matches()) {
                    String varName = matcher.group(1);
                    String varValue = matcher.group(2);
                    envVars.put(varName, varValue);
                } else {
                    throw new IOException("Invalid format in " + ENV_FILE + " at line " + lineNum + ": " + line);
                }
            }
        }

        String[] requiredVars = {
                "CAMUNDA_CLUSTER_ID",
                "CAMUNDA_CLIENT_ID",
                "CAMUNDA_CLIENT_SECRET",
                "CAMUNDA_CLUSTER_REGION"
        };

        for (String var : requiredVars) {
            if (!envVars.containsKey(var)) {
                throw new IOException("Missing required environment variable: " + var);
            }
        }

        return envVars;
    }

    private static void testConnection(Map<String, String> envVars) throws Exception {
        String clusterId = envVars.get("CAMUNDA_CLUSTER_ID");
        String clientId = envVars.get("CAMUNDA_CLIENT_ID");
        String clientSecret = envVars.get("CAMUNDA_CLIENT_SECRET");
        String region = envVars.get("CAMUNDA_CLUSTER_REGION");

        System.out.println("Using client ID: " + maskCredential(clientId));

        try (CamundaClient client = CamundaClient.newCloudClientBuilder()
                .withClusterId(clusterId)
                .withClientId(clientId)
                .withClientSecret(clientSecret)
                .withRegion(region)
                .build()) {

            System.out.println("Testing connection...");
            Topology topology = client.newTopologyRequest().send().join();

            if (topology.getBrokers().isEmpty()) {
                throw new Exception("Connected but no brokers found in topology");
            }

        } catch (CompletionException e) {
            throw (Exception) e.getCause();
        }
    }

    private static String maskCredential(String credential) {
        if (credential == null || credential.length() < 8) {
            return "***";
        }
        return credential.substring(0, 4) + "****" + credential.substring(credential.length() - 4);
    }

    private static void handleException(Exception e) {
        String errorMsg = e.getMessage() != null ? e.getMessage().toLowerCase() : "";
        String className = e.getClass().getName().toLowerCase();

        if (errorMsg.contains("ssl") || errorMsg.contains("certificate") || className.contains("ssl")) {
            printError("SSL error: " + e.getMessage());
            System.exit(EXIT_SSL_ERROR);
        } else if (errorMsg.contains("connection") || errorMsg.contains("connect") ||
                errorMsg.contains("network") || errorMsg.contains("timeout") ||
                errorMsg.contains("timed out") || className.contains("connection") ||
                className.contains("timeout")) {
            printError("Connection error: " + e.getMessage());
            System.exit(EXIT_CONNECTION_ERROR);
        } else if (errorMsg.contains("401") || errorMsg.contains("403") ||
                errorMsg.contains("unauthorized") || errorMsg.contains("forbidden") ||
                errorMsg.contains("authentication") || errorMsg.contains("token") ||
                errorMsg.contains("credential")) {
            printError("Authentication error: " + e.getMessage());
            System.exit(EXIT_AUTH_ERROR);
        } else {
            printError("Unexpected error: " + e.getMessage());
            if (e.getCause() != null) {
                System.err.println("Caused by: " + e.getCause().getMessage());
            }
            System.exit(EXIT_OTHER_ERROR);
        }
    }

    private static void printError(String message) {
        System.out.println("***** CONNECTION FAILED: " + message + " *****");
    }

    private static void printSuccess() {
        System.out.println("***** CONNECTION SUCCESSFUL *****");
    }
}

File: python/testConnection.py
#!/usr/bin/env python3
"""
Camunda 8 SaaS Connection Test Tool - Python Version

This tool tests connectivity to Camunda 8 SaaS by:
1. Checking Python version (must be 3.8+)
2. Loading environment variables from envVars.txt
3. Obtaining OAuth token
4. Testing API connection
5. Validating response content

All dependencies are bundled in the lib/ directory.

Author: Camunda Academy
"""

import sys
import os
import json
import re
from typing import Dict, Tuple

# Add bundled libraries to path
lib_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'lib')
if os.path.exists(lib_path) and lib_path not in sys.path:
    sys.path.insert(0, lib_path)

# Exit codes
EXIT_SUCCESS = 0
EXIT_SSL_ERROR = 1
EXIT_CONNECTION_ERROR = 2
EXIT_AUTH_ERROR = 3
EXIT_OTHER_ERROR = 4

# Import requests from bundled libraries
try:
    import requests
except ImportError:
    print("***** ERROR: Required 'requests' library not found *****")
    print("This package should include bundled dependencies.")
    print("Please contact your training manager.")
    sys.exit(EXIT_OTHER_ERROR)


def check_python_version() -> None:
    version_info = sys.version_info
    if version_info < (3, 8):
        print_error(
            f"Python 3.8 or higher is required. Current version: {version_info.major}.{version_info.minor}.{version_info.micro}\n"
            "Please upgrade Python or contact your training manager.\n"
            "Download Python from: https://www.python.org/downloads/"
        )
        sys.exit(EXIT_OTHER_ERROR)
    print(f"Python version check passed: {version_info.major}.{version_info.minor}.{version_info.micro}")


def print_error(message: str) -> None:
    print(f"***** CONNECTION FAILED: {message} *****")


def print_success() -> None:
    print("***** CONNECTION SUCCESSFUL *****")


def load_env_vars(file_path: str = "envVars.txt") -> Dict[str, str]:
    if not os.path.isfile(file_path):
        parent_file_path = os.path.join("..", file_path)
        if os.path.isfile(parent_file_path):
            file_path = parent_file_path
        else:
            raise FileNotFoundError(f"{file_path} file not found. Double check that this file is available in this directory or parent directory")
    
    env_vars = {}
    
    try:
        with open(file_path, 'r') as f:
            for line_num, line in enumerate(f, 1):
                line = line.strip()
                if not line or line.startswith('#'):
                    continue
                
                match = re.match(r"export\s+(\w+)='(.+)'", line)
                if match:
                    var_name, var_value = match.groups()
                    env_vars[var_name] = var_value
                else:
                    raise ValueError(f"Invalid format in {file_path} at line {line_num}: {line}")
    
    except Exception as e:
        raise ValueError(f"Error reading {file_path}: {str(e)}")
    
    required_vars = [
        'CAMUNDA_CONSOLE_CLIENT_ID',
        'CAMUNDA_CONSOLE_CLIENT_SECRET',
        'CAMUNDA_OAUTH_URL',
        'CAMUNDA_CONSOLE_BASE_URL',
        'CAMUNDA_CONSOLE_OAUTH_AUDIENCE'
    ]
    
    missing_vars = [var for var in required_vars if var not in env_vars]
    if missing_vars:
        raise ValueError(f"Missing required environment variables: {', '.join(missing_vars)}")
    
    return env_vars


def get_access_token(env_vars: Dict[str, str]) -> str:
    token_request_data = {
        "grant_type": "client_credentials",
        "audience": env_vars['CAMUNDA_CONSOLE_OAUTH_AUDIENCE'],
        "client_id": env_vars['CAMUNDA_CONSOLE_CLIENT_ID'],
        "client_secret": env_vars['CAMUNDA_CONSOLE_CLIENT_SECRET']
    }
    
    headers = {
        "Content-Type": "application/json"
    }
    
    try:
        response = requests.post(
            env_vars['CAMUNDA_OAUTH_URL'],
            json=token_request_data,
            headers=headers,
            timeout=30
        )
        response.raise_for_status()
        
        token_response = response.json()
        access_token = token_response.get('access_token')
        
        if not access_token:
            raise ValueError("Access token not found in response")
        
        return access_token
        
    except Exception as e:
        if hasattr(e, '__class__') and 'requests' in str(e.__class__):
            if 'SSL' in str(e):
                raise Exception(f"SSL certificate error: {str(e)}")
            elif 'Connection' in str(e):
                raise Exception(f"Connection error: {str(e)}")
            elif 'Timeout' in str(e):
                raise Exception(f"Request timeout: {str(e)}")
            elif 'HTTP' in str(e):
                raise Exception(f"HTTP error {response.status_code}: {response.text}")
            else:
                raise Exception(f"Request error: {str(e)}")
        elif isinstance(e, json.JSONDecodeError):
            raise Exception(f"Invalid JSON response: {str(e)}")
        else:
            raise e


def test_api_connection(env_vars: Dict[str, str], access_token: str) -> Tuple[int, str]:
    headers = {
        "Authorization": f"Bearer {access_token}"
    }
    
    api_url = f"{env_vars['CAMUNDA_CONSOLE_BASE_URL']}/members"
    
    try:
        response = requests.get(
            api_url,
            headers=headers,
            timeout=30
        )
        
        return response.status_code, response.text
        
    except Exception as e:
        if hasattr(e, '__class__') and 'requests' in str(e.__class__):
            if 'SSL' in str(e):
                raise Exception(f"SSL certificate error: {str(e)}")
            elif 'Connection' in str(e):
                raise Exception(f"Connection error: {str(e)}")
            elif 'Timeout' in str(e):
                raise Exception(f"Request timeout: {str(e)}")
            else:
                raise Exception(f"Request error: {str(e)}")
        else:
            raise e


def validate_response(status_code: int, response_body: str) -> bool:
    if status_code != 200:
        print_error(f"Status code: {status_code} Response body: {response_body}")
        return False
    
    if '"name"' in response_body and '"email"' in response_body:
        return True
    else:
        print_error("Response does not contain required attributes.")
        return False


def main() -> None:
    try:
        check_python_version()
        
        print("Loading environment variables...")
        env_vars = load_env_vars()
        
        print("Requesting access token...")
        access_token = get_access_token(env_vars)
        
        print("Testing API connection...")
        status_code, response_body = test_api_connection(env_vars, access_token)
        
        if validate_response(status_code, response_body):
            print_success()
            sys.exit(EXIT_SUCCESS)
        else:
            sys.exit(EXIT_OTHER_ERROR)
        
    except FileNotFoundError as e:
        print_error(str(e))
        sys.exit(EXIT_OTHER_ERROR)
    except ValueError as e:
        error_msg = str(e).lower()
        if "token" in error_msg or "access_token" in error_msg or "authentication" in error_msg or "unauthorized" in error_msg:
            print_error(f"Authentication failed: {str(e)}")
            sys.exit(EXIT_AUTH_ERROR)
        else:
            print_error(str(e))
            sys.exit(EXIT_OTHER_ERROR)
    except Exception as e:
        error_msg = str(e).lower()
        
        if "ssl" in error_msg or "certificate" in error_msg:
            print_error(f"SSL error: {str(e)}")
            sys.exit(EXIT_SSL_ERROR)
        elif "connection" in error_msg or "proxy" in error_msg or "network" in error_msg:
            print_error(f"Connection error: {str(e)}")
            sys.exit(EXIT_CONNECTION_ERROR)
        elif "timeout" in error_msg or "timed out" in error_msg:
            print_error(f"Connection error (timeout): {str(e)}")
            sys.exit(EXIT_CONNECTION_ERROR)
        elif "401" in error_msg or "403" in error_msg or "unauthorized" in error_msg or "forbidden" in error_msg or "authentication" in error_msg or "token" in error_msg:
            print_error(f"Authentication error: {str(e)}")
            sys.exit(EXIT_AUTH_ERROR)
        else:
            print_error(f"Unexpected error: {str(e)}")
            sys.exit(EXIT_OTHER_ERROR)


if __name__ == "__main__":
    main()